kong/db/declarative/import.lua

```diff
@@ -203,6 +258,122 @@ local function config_is_empty(entities)
 end
 
 
+-- common implementation for
+-- insert_entity_for_txn() and delete_entity_for_txn()
+local function _set_entity_for_txn(t, entity_name, item, options, is_delete)
+  local dao = kong.db[entity_name]
+  local schema = dao.schema
+  local pk = pk_string(schema, item)
+  local ws_id = workspace_id(schema, options)
+
+  local itm_key = item_key(entity_name, ws_id, pk)
+
+  -- if we are deleting, item_value and idx_value should be nil
+  local itm_value, idx_value
+
+  -- if we are inserting or updating
+  -- itm_value is serialized entity
+  -- idx_value is the lmdb item_key
+  if not is_delete then
+    local err
+
+    -- serialize item with possible nulls
+    itm_value, err = marshall(item)
+    if not itm_value then
+      return nil, err
+    end
+
+    idx_value = itm_key
+  end
+
+  -- store serialized entity into lmdb
+  t:set(itm_key, itm_value)
+
+  -- select_by_cache_key
+  if schema.cache_key then
+    local cache_key = dao:cache_key(item)
+    local key = unique_field_key(entity_name, ws_id, "cache_key", cache_key)
+
+    -- store item_key or nil into lmdb
+    t:set(key, idx_value)
+  end
+
+  for fname, fdata in schema:each_field() do
+    local is_foreign = fdata.type == "foreign"
+    local fdata_reference = fdata.reference
+    local value = item[fname]
+
+    -- value may be null, we should skip it
+    if not value or value == null then
+      goto continue
+    end
+
+    -- value should be a string or table
+
+    local value_str
+
+    if fdata.unique then
+      -- unique and not a foreign key, or is a foreign key, but non-composite
+      -- see: validate_foreign_key_is_single_primary_key, composite foreign
+      -- key is currently unsupported by the DAO
+      if type(value) == "table" then
+        assert(is_foreign)
```

Comment:
This `assert` is for a bug scenario that should never happen. If it did happen it is better to let the entire call chain fail instead of trying to save it.